### PR TITLE
[ fix #729 ] List membership irrelevance

### LINKED ISSUE #729: With K, `x ∈ xs` is irrelevant when `Unique xs`
Worth proving under the assumption that K holds for the underlying set,
and maybe provide without the precondition in a `.WithK` module.

### PR DESCRIPTION
**Pull request by iitalics**
closes #729

### CODE REVIEW

**Comment by iitalics** [comment]
This lemma, `∉×∈⇒≉ : ∀ {x y xs} → All (y ≉_) xs → x ∈ xs → x ≉ y` could have been derived from a more general version of [lookup](https://agda.github.io/agda-stdlib/Data.List.Relation.Unary.All.html#1513):
```agda
module _ {p q} {P : Pred A p} {Q : Pred A q}
         (resp : P ⟶ Q Respects _≈_) where
  lookup′ : ∀ {x xs} → All P xs → x ∈ xs → Q x
  lookup′ (py ∷ pxs) (here x≈y)   = resp (sym x≈y) py
  lookup′ (py ∷ pxs) (there x∈xs) = lookup′ pxs x∈xs

∉×∈⇒≉′ : ∀ {x y xs} → All (y ≉_) xs → x ∈ xs → x ≉ y
∉×∈⇒≉′ = lookup′ λ x≈z y≉x z≈y → y≉x (sym (trans x≈z z≈y))
```

I couldn't find anything existing like that, should it be added?

**Comment by gallais** [comment]
In sublist we have a [similar generalised lookup](https://github.com/agda/agda-stdlib/blob/master/src/Data/List/Relation/Binary/Sublist/Heterogeneous.agda#L56) so I'd say yes it would be nice addition
(using `Any` rather than the `x ∈ xs` special case).

File: src/Data/List/Membership/Setoid/Properties.agda [hunk 1]
```diff
@@ -15,14 +15,17 @@ open import Data.List.Relation.Unary.Any as Any using (Any; here; there)
...
 open import Data.Nat.Properties using (≤-trans; n≤1+n)
 open import Data.Product as Prod using (∃; _×_; _,_ ; ∃₂)
 open import Data.Sum using (_⊎_; inj₁; inj₂)
 open import Function using (_$_; flip; _∘_; id)
 open import Relation.Binary hiding (Decidable)
+  renaming (Irrelevant to Irrelevant₂)
 open import Relation.Binary.PropositionalEquality as P using (_≡_)
 open import Relation.Unary using (Decidable; Pred)
+  renaming (Irrelevant to Irrelevant₁)
 open import Relation.Nullary using (¬_; yes; no)
```
Comment (gallais) [inline src/Data/List/Membership/Setoid/Properties.agda]: Regarding name clashes, we typicall do:

```agda
open import Relation.Binary as B (hiding / using / etc.)
open import Relation.Unary as U (hiding / using / etc.)
```

And then use qualified names to disambiguate: `U.Irrelevant` vs. `B.Irrelevant`. This way we
don't need to have long renaming lists when there are a lot of clashes.

**Comment by iitalics** [comment]
What do you think of [a4cae25](https://github.com/agda/agda-stdlib/pull/766/commits/a4cae257462be3340d76e72d6870659b254e4ccd)?

File: src/Data/List/Relation/Unary/All.agda [hunk 2]
```diff
@@ -135,6 +130,21 @@ module _ p {A : Set a} {P : Pred A (a ⊔ p)}
   forM : ∀ {Q : Pred A q} {xs} → All Q xs → (Q ⊆ M ∘′ P) → M (All P xs)
   forM = forA p App
 
+------------------------------------------------------------------------
+-- Generalised lookup based on a proof of Any
+
+module _ {P : Pred A p} {Q : Pred A q} {R : Pred A r}
+         (p×q⇒r : ∀[ P ⇒ Q ⇒ R ]) where
```
Comment (MatthewDaggitt) [inline src/Data/List/Relation/Unary/All.agda]: Please could we have `p×q⇒r : ∀[ P ⇒ Q ⇒ R ]` in the main type as it's quite an important part of it?

File: src/Data/List/Membership/Setoid/Properties.agda [hunk 3]
```diff
@@ -50,6 +52,28 @@ module _ {c ℓ} (S : Setoid c ℓ) where
...
+  open Unique S
+  open Membership S
+
+  private
+    _≉_ = λ x y → ¬ (x ≈ y)
+
+    ∉×∈⇒≉ : ∀ {x y xs} → All (y ≉_) xs → x ∈ xs → x ≉ y
+    ∉×∈⇒≉ = All.glookup λ y≉z x≈z x≈y → y≉z (trans (sym x≈y) x≈z)
+
+  irrelevant : ∀ {xs} → Unique xs → U.Irrelevant (_∈ xs)
```
Comment (MatthewDaggitt) [inline src/Data/List/Membership/Setoid/Properties.agda]: Do you think we should rename the lemmas `unique⇒irrelevant`? The uniqueness is quite crucial.

Comment (MatthewDaggitt) [reply src/Data/List/Membership/Setoid/Properties.agda]: Also the setoid version should be added to the CHANGELOG.

File: src/Data/List/Membership/Setoid/Properties.agda [hunk 4]
```diff
@@ -50,6 +52,28 @@ module _ {c ℓ} (S : Setoid c ℓ) where
   ∉-resp-≋ : ∀ {x} → (x ∉_) Respects _≋_
   ∉-resp-≋ xs≋ys v∉xs v∈ys = v∉xs (∈-resp-≋ (≋-sym xs≋ys) v∈ys)
 
+------------------------------------------------------------------------
+-- Irrelevance
+
+module _ {c ℓ} (S : Setoid c ℓ)
+               (≈-irrelevant : B.Irrelevant (Setoid._≈_ S)) where
```
Comment (MatthewDaggitt) [inline src/Data/List/Membership/Setoid/Properties.agda]: Again the irrelevance of the equality is important, could be moved to the proof? Apologies, I really should have spotted this before!

Comment (iitalics) [reply src/Data/List/Membership/Setoid/Properties.agda]: :+1:

**Comment by MatthewDaggitt** [comment]
Great, thanks! I'll leave it up a few more days and then merge in.

File: src/Data/List/Membership/Setoid/Properties.agda [hunk 5]
```diff
@@ -50,6 +52,31 @@ module _ {c ℓ} (S : Setoid c ℓ) where
...
+
+module _ {c ℓ} (S : Setoid c ℓ) where
+
+  open Setoid S
+  open Unique S
+  open Membership S
+
+  private
+    _≉_ = λ x y → ¬ (x ≈ y)
+
```
Comment (gallais) [inline src/Data/List/Membership/Setoid/Properties.agda]: @MatthewDaggitt Would it make sense to add this definition to `Setoid`?

Comment (MatthewDaggitt) [reply src/Data/List/Membership/Setoid/Properties.agda]: Yup, I included it in `IndexedSetoid`.

**Comment by gallais** [comment]
I'm happy with the current state the PR is in.

I'm not a huge fan of the name `glookup` though. It seems to me we could call this
function `lookupWith` by analogy with `zipWith`: the more direct definition would
return a pair of proofs of `P` and `Q` but we somehow combine them using the
functional argument.

**Comment by MatthewDaggitt** [comment]
@gallais `_≉_` is also defined a couple of times in `Data.List.Relation.Unary.Unique.Setoid.Properties`.

**Comment by gallais** [comment]
Nice. Thanks @iitalics!

**Pull request merged**